### PR TITLE
fix: Resolve null pointer problem when metadata undefined

### PR DESCRIPTION
**Pull request by abyss-w**
fix: Resolve null pointer problem when metadata undefined

### CODE REVIEW

**Comment by Xuanwo** [comment]
Please add a comment or some thing to trigger the integration tests.

File: storage.go [hunk 1]
```diff
@@ -620,18 +620,20 @@ func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o
...
-		// The path is a symlink object.
-		if !s.features.VirtualLink {
-			err = NewOperationNotImplementedError("virtual_link")
-			return nil, err
-		}
+	if output.Metadata != nil {
+		metadata := output.Metadata
+		if target, ok := metadata["X-Amz-Meta-Bs-Link-Target"]; ok {
+			// The path is a symlink object.
+			if !s.features.VirtualLink {
```
Comment (Xuanwo) [inline storage.go:627]: If virtual link is not enabled, we should treat it as a normal object instead return error directly.

File: storage.go [hunk 2]
```diff
@@ -620,18 +620,20 @@ func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o
...
-	metadata := output.Metadata
-	if target, ok := metadata["x-amz-meta-bs-link-target"]; ok {
-		// The path is a symlink object.
-		if !s.features.VirtualLink {
-			err = NewOperationNotImplementedError("virtual_link")
-			return nil, err
-		}
+	if output.Metadata != nil {
+		metadata := output.Metadata
+		if target, ok := metadata["X-Amz-Meta-Bs-Link-Target"]; ok {
```
Comment (Xuanwo) [inline storage.go:625]: Maybe we can use it as a const?

File: storage.go (hunk 1 above)
Comment (abyss-w) [reply storage.go:627]: Got it!

File: storage.go (hunk 2 above)
Comment (abyss-w) [reply storage.go:625]: The situation here is the same as that of qingstor.

Comment (Xuanwo) [reply storage.go:625]: Hi, please read s3 sdk's doc.

```go
	// By default unmarshaled keys are written as a map keys in following canonicalized format:
	// the first letter and any letter following a hyphen will be capitalized, and the rest as lowercase.
	// Set `aws.Config.LowerCaseHeaderMaps` to `true` to write unmarshaled keys to the map as lowercase.
```

I think both input and output should in lowercase.

Comment (abyss-w) [reply storage.go:625]: OK, got it!

File: storage.go (hunk 1 above)
Comment (abyss-w) [reply storage.go:627]: > If virtual link is not enabled, we should treat it as a normal object instead return error directly.

I'm a bit confused, what is the mode of this normal object set to?
Or even if you don't have virtual link enabled, you can still set it to link mode?

Comment (JinnyYi) [reply storage.go:627]: Virtual link not enabled means no simulated symlink for the service, so all the object mode should not be `ModeLink`.  Treat as a normal object means we should continue the assignment for the returning object here and check whether the object is a nomal object(ModeRead) or a simulated dir(ModeDir).

Comment (abyss-w) [reply storage.go:627]: > Virtual link not enabled means no simulated symlink for the service, so all the object mode should not be `ModeLink`. Treat as a normal object means we should continue the assignment for the returning object here and check whether the object is a nomal object(ModeRead) or a simulated dir(ModeDir).

Got it!

File: storage.go [hunk 3]
```diff
@@ -132,12 +132,10 @@ func (s *Storage) createDir(ctx context.Context, path string, opt pairStorageCre
 	return o, nil
 }
 
-func (s *Storage) createLink(ctx context.Context, path string, target string, opt pairStorageCreateLink) (o *Object, err error) {
-	if !s.features.VirtualLink {
-		err = NewOperationNotImplementedError("virtual_link")
-		return nil, err
-	}
+// LinkMeta is the name of the user-defined metadata name used to store the target.
+const LinkMeta = "x-amz-meta-bs-link-target"
```
Comment (Xuanwo) [inline storage.go:136]: How about keeping this header key private?

How about names like `metadataLinkTargetHeader`?

Comment (abyss-w) [reply storage.go:136]: Got it!

**Comment by Xuanwo** [comment]
Great work!

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**